onnxruntime/core/language_interop_ops/torch/torch_proxy.cc

```diff
@@ -0,0 +1,361 @@
+
+  for (const auto i : tensor_indices) {
+    ORT_ENFORCE(i >= 0 && static_cast<size_t>(i) < len, "Index range is from 0 to ", len - 1, ", but found ", i);
+    counts.at(i) += 1;
+  }
+  for (const auto i : obj_indices) {
+    ORT_ENFORCE(i >= 0 && static_cast<size_t>(i) < len, "Index range is from 0 to ", len - 1, ", but found ", i);
+    counts.at(i) += 1;
+  }
+  for (size_t i = 0; i < len; ++i) {
+    ORT_ENFORCE(counts.at(i) == 1, "Duplicated or unused argument index detected: ", i);
+  }
+#endif
+}
+
+// len: the number of input arguments.
+// tensor_indices: if tensor_indices[i] is j,
+//                 then the j-th input argument should be a tensor.
+PyObject* CreateTensorFlags(
+    const size_t len,
+    const std::vector<int64_t>& tensor_indices) {
+  PyObject* flags = Ort_PyList_New(len, "tensor_flags_list");
+
+  // First we fill the list with 0. Later we will
+  // assign 1's to tensors' corresponding positions.
+  for (size_t i = 0; i < len; ++i) {
+    PyObject* zero = PyLong_FromLong(0);
+    Ort_PyList_SetItem_NoIncref(flags, i, zero, std::to_string(__LINE__));
+  }
+
+  for (const auto i : tensor_indices) {
+    PyObject* one = PyLong_FromLong(1);
+    Ort_PyList_SetItem_NoIncref(flags, i, one, std::to_string(__LINE__));
+  }
+
+  return flags;
+}
+
+// flags[i] corresponds to the i-th input of apply/backward.
+PyObject* CreateRequiresGradFlags(
+    const std::vector<int64_t>& requires_grads) {
+  PyObject* flags = Ort_PyList_New(requires_grads.size(), "require_grads_list");
+  for (size_t i = 0; i < requires_grads.size(); ++i) {
+    PyObject* value;
+    if (requires_grads.at(i) != 0) {
+      value = PyLong_FromLong(1);
+    } else {
+      value = PyLong_FromLong(0);
+    }
+    Ort_PyList_SetItem_NoIncref(flags, i, value, std::to_string(__LINE__));
+  }
+  return flags;
+}
+
+void InvokeRunner(
+    PyObject* callback_runner,
+    PyObject* args,
+    bool is_training_mode,
+    void** diff_ctx,
+    std::vector<OrtValue>& returned_ortvalues) {
+  PythonObjectPtr result_ptr(PyObject_CallObject(callback_runner, args));
+
+  if (PyErr_Occurred()) {
+    PyErr_Print();
+    ORT_THROW("Python function execution fails with the above information.");
+  }
+
+  ORT_ENFORCE(PyTuple_Check(result_ptr.get()), "Python function must return a tuple.");
+
+  size_t i = 0;
+  if (diff_ctx) {
+    // Assume that the first input element in the returned tuple is autograd context
+    // from Pytorch.
+    PyObject* py_obj = PyTuple_GetItem(result_ptr.get(), 0);
+    if (is_training_mode) {
+      const auto& refcnt = Py_REFCNT(py_obj);
+      // we don't need do ref increase here because, python return tensor.grad_fn as part of
+      // tuple, who increased the refcnt already (and tensor persist until the backward kernels completed).
+      // Pytorch also increases refcnt before apply() return, so we should expect refcount = 2 here.
+      ORT_ENFORCE(refcnt == 2, "Ref count of context should be 2, but actually it's ", refcnt, ".");
```

Comment:
Can we rely on == 2, as opposed to >= 2?  The case I'm worried about is that the user's code (for some reason unknown to us) might stash its own copies of the context object?

Reply:
yeah, this is possible when they for example store the ctx in a global variable in forward function body, and release it in backward function body. Let me add a >=2, and besidesI would prefer to add a warning log here, if we find a>2, make sense?